Userland/Libraries/LibVideo/Containers/Matroska/Reader.cpp

```diff
@@ -98,35 +103,65 @@ DecoderErrorOr<Reader> Reader::from_data(ReadonlyBytes data)
     return reader;
 }
 
-static DecoderErrorOr<void> parse_master_element(Streamer& streamer, [[maybe_unused]] StringView element_name, Function<DecoderErrorOr<IterationDecision>(u64, size_t)> element_consumer)
+// Returns the position of the first element that is read from this master element.
+static DecoderErrorOr<size_t> parse_master_element(Streamer& streamer, [[maybe_unused]] StringView element_name, Function<DecoderErrorOr<IterationDecision>(u64)> element_consumer)
 {
     auto element_data_size = TRY_READ(streamer.read_variable_size_integer());
     dbgln_if(MATROSKA_DEBUG, "{} has {} octets of data.", element_name, element_data_size);
 
+    bool first_element = true;
+    auto first_element_position = streamer.position();
+
     streamer.push_octets_read();
     while (streamer.octets_read() < element_data_size) {
         dbgln_if(MATROSKA_TRACE_DEBUG, "====== Reading  element ======");
         auto element_id = TRY_READ(streamer.read_variable_size_integer(false));
-        auto element_position = streamer.position();
         dbgln_if(MATROSKA_TRACE_DEBUG, "{:s} element ID is {:#010x}", element_name, element_id);
 
-        auto result = element_consumer(element_id, element_position);
+        if (element_id == EBML_CRC32_ELEMENT_ID) {
+            // The CRC-32 Element contains a 32-bit Cyclic Redundancy Check value of all the
+            // Element Data of the Parent Element as stored except for the CRC-32 Element itself.
+            // When the CRC-32 Element is present, the CRC-32 Element MUST be the first ordered
+            // EBML Element within its Parent Element for easier reading.
+            if (!first_element)
+                return DecoderError::corrupted("CRC32 element must be the first child"sv);
+
+            // All Top-Level Elements of an EBML Document that are Master Elements SHOULD include a
+            // CRC-32 Element as a Child Element. The CRC in use is the IEEE-CRC-32 algorithm as used
+            // in the [ISO3309] standard and in Section 8.1.1.6.2 of [ITU.V42], with initial value of
+            // 0xFFFFFFFF. The CRC value MUST be computed on a little-endian bytestream and MUST use
+            // little-endian storage.
+
+            // NOTE: We skip the CRC32 element instead of checking it. This could be changed
```

Comment:
Seem worth adding a FIXME to verify the CRC?

Reply:
Good point, I've replaced this with a FIXME.

I added a note that were we to implement the checks, it would probably be best to only verify metadata in the container, and not the data sent to the codec. If we verify the Cluster Elements' contents, that would require us to buffer an entire cluster at once and then run the checksum, when we can begin pulling samples as soon as one is available instead. It would also prevent the codecs from attempting error correction when possible.

I also generally prefer not checking integrity so we can see more bugs :^) as a later thing, though, it may be useful